clippy_lints/src/loops.rs

```diff
@@ -2358,17 +2358,100 @@ fn check_infinite_loop<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, cond: &'tcx Expr, e
         return;
     };
     let mutable_static_in_cond = var_visitor.def_ids.iter().any(|(_, v)| *v);
-    if no_cond_variable_mutated && !mutable_static_in_cond {
+    let has_break_or_return = if let ExprKind::Block(ref block, _) = expr.kind {
+        match infinite_loop_block(block) {
+            InfiniteLoopResult::MayBreak => true,
+            InfiniteLoopResult::Otherwise => false,
+        }
+    } else {
+        panic!("Expected block here always!");
+    };
+    if no_cond_variable_mutated && !mutable_static_in_cond && !has_break_or_return {
         span_lint(
             cx,
             WHILE_IMMUTABLE_CONDITION,
             cond.span,
-            "Variable in the condition are not mutated in the loop body. \
-             This either leads to an infinite or to a never running loop.",
+            "variables in the condition are not mutated in the loop body. \
+             This may lead to an infinite or to a never running loop",
         );
     }
 }
 
+// inspired by never_loop_block()
+fn infinite_loop_block(block: &Block) -> InfiniteLoopResult {
+    let stmts = block.stmts.iter().map(stmt_to_expr);
+    let expr = once(block.expr.as_ref().map(|p| &**p));
+    let mut iter = stmts.chain(expr).filter_map(|e| e);
+    infinite_loop_expr_seq(&mut iter)
+}
+
+enum InfiniteLoopResult {
+    // Has break/return in the body of the loop
+    MayBreak,
+    Otherwise,
+}
+
+fn infinite_loop_expr_seq<'a, T: Iterator<Item = &'a Expr>>(es: &mut T) -> InfiniteLoopResult {
+    es.map(|e| infinite_loop_expr(e))
+        .fold(InfiniteLoopResult::Otherwise, infinite_loop_combine_seq)
```

Comment:
This is more efficient:
```suggestion
        .find(|res| res == InfiniteLoopResult::MayBreak)
        .unwrap_or(InfiniteLoopResult::Otherwise)
```

(Requires `PartialEq` derive on the `InfiniteLoopResult` enum)